File: Programmers/cpp/LEVEL_3_야근_지수.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

long long solution(int n, vector<int> works) {
	long long answer = 0;
	priority_queue<int> hi;

	for (int n : works) {
		hi.push(n);
	}

	while (n > 0) {
		if (hi.empty()) {
			break;
		}

		int num = hi.top();
		hi.pop();

		num--;
		if (num > 0) {
			hi.push(num);
		}

		n--;
	}

	while (!hi.empty()) {
		answer += (hi.top() * hi.top());
		hi.pop();
	}

	return answer;
}

File: Programmers/java/LEVEL_3_야근_지수.java
import java.util.*;

class Solution {
    public long solution(int n, int[] works) {
        long answer = 0;
        PriorityQueue<Integer> hi = new PriorityQueue<>(Collections.reverseOrder());

        for (int num : works) {
            hi.add(num);
        }

        while (n > 0) {
            if (hi.isEmpty()) {
                break;
            }

            int num = hi.peek();
            hi.poll();

            num--;
            if (num > 0) {
                hi.add(num);
            }

            n--;
        }

        while (!hi.isEmpty()) {
            answer += (hi.peek() * hi.peek());
            hi.poll();
        }

        return answer;
    }
}